### PR TITLE
os/bluestore: speed up pool removal by introducing collection list prefetch

### PR DESCRIPTION
**Pull request by ifed01**
Also optionally retrieves onode during collection listing and puts it into the cache to reuse during removal.
Signed-off-by: Igor Fedotov <ifedotov@suse,com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by neha-ojha** [comment]
@liewegas can you take a look at this?
@ifed01 needs rebase

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -10145,17 +10252,18 @@ int BlueStore::_upgrade_super()
   return 0;
 }
 
-void BlueStore::_assign_nid(TransContext *txc, OnodeRef o)
+bool BlueStore::_assign_nid(TransContext *txc, OnodeRef o)
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: Instead of using the _assign_nid() return value to determine if a key is new, I suggest putting it directly in get_onode() in the create path, i.e. with
```
    // new object, new onode
    on = new Onode(this, oid, key);
```
that way we have just (1) the normal create path (there), (2) the remove case, and (3) the wonky clone and rename cases.

File: src/osd/PG.cc [hunk 2]
```diff
@@ -3776,7 +3776,7 @@ void PG::do_delete_work(ObjectStore::Transaction &t)
     max,
     &olist,
     &next,
-    0);
+    CEPH_OSD_OP_FLAG_FADVISE_WILLNEED | CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL);
```
Comment (liewegas) [inline src/osd/PG.cc]: put this in a separate patch?

also, should update the scrub caller with both of these flags...

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: done

File: src/osd/PG.cc (hunk 2 above)
Comment (ifed01) [reply src/osd/PG.cc]: > put this in a separate patch?
done
> 
> also, should update the scrub caller with both of these flags...
not sure if this makes sense - currently prefetch is triggered for listings started from the beginning only.

**Comment by ifed01** [comment]
@aclamk - rebased, ready for review

**Comment by tchaikov** [comment]
@aclamk ping?

**Comment by ifed01** [comment]
@tchaikov, @aclamk - I think this is a legacy PR now. Currently one is able to use larger collection listings and partially apply them. Hence no need for a caching inside BlueStore. Working on a competing PG now. And will close this one once new PR is ready.
